src/RobotLoader/RobotLoader.php

```diff
@@ -347,43 +344,92 @@ public function setAutoRefresh($on = TRUE)
 
 
 	/**
+	 * Sets path to temporary directory.
 	 * @return self
 	 */
-	public function setCacheStorage(Nette\Caching\IStorage $storage)
+	public function setTempDirectory($path)
 	{
-		$this->cacheStorage = $storage;
+		$this->tempDirectory = $path;
 		return $this;
 	}
 
 
 	/**
-	 * @return Nette\Caching\IStorage
+	 * @return void
 	 */
-	public function getCacheStorage()
+	private function loadCache()
 	{
-		return $this->cacheStorage;
+		if (!$this->tempDirectory) {
+			return;
+		}
+		$file = $this->getCacheFile();
+		$list = @include $file;
+		if (!is_array($list)) {
+			$handle = fopen("$file.lock", 'c+');
+			if (!$handle || !flock($handle, LOCK_EX)) {
+				throw new \RuntimeException("Unable to acquire exclusive lock '$file.lock'.");
+			}
+
+			$list = @include $file;
+			if (!is_array($list)) {
+				$this->refreshCache();
+				$this->saveCache();
+				$list = $this->classes;
+			}
+
+			flock($handle, LOCK_UN);
+			fclose($handle);
+			@unlink("$file.lock"); // @ file may become locked on Windows
+		}
+		$this->classes = $list;
 	}
 
 
 	/**
-	 * @return Nette\Caching\Cache
+	 * @return void
 	 */
-	protected function getCache()
+	private function saveCache()
 	{
-		if (!$this->cacheStorage) {
-			trigger_error('Missing cache storage.', E_USER_WARNING);
-			$this->cacheStorage = new Nette\Caching\Storages\DevNullStorage;
+		if (!$this->tempDirectory) {
+			return;
+		} elseif (!is_dir($this->tempDirectory)) {
+			@mkdir($this->tempDirectory); // @ - directory may already exist
+		}
+
+		$file = $this->getCacheFile();
+		$code = '<?php return ' . var_export($this->classes, TRUE) . ';';
+		if (file_put_contents("$file.tmp", $code) !== strlen($code) || !rename("$file.tmp", $file)) {
+			@unlink("$file.tmp"); // @ - file may not exist
+			throw new \RuntimeException("Unable to create '$file'.");
+		} elseif (function_exists('opcache_invalidate')) {
+			@opcache_invalidate($file, TRUE); // @ can be restricted
 		}
-		return new Cache($this->cacheStorage, 'Nette.RobotLoader');
 	}
 
 
 	/**
 	 * @return string
 	 */
-	protected function getKey()
+	private function getCacheFile()
+	{
+		return $this->tempDirectory . '/' . md5(serialize($this->getCacheKey())) . '.php';
+	}
+
+
+	/**
+	 * @return string
```

Comment:
should be `@return array` (or `@return mixed` in theory, as it may be overridden to return anything that is serializable, but i like `@return array` more.